### PR TITLE
FEAT-1402: LAZ reader

### PR DESCRIPTION
**Pull request by leo-stan**

### CODE REVIEW

**Comment by leo-stan** [comment]
Done by Chris

File: cpp/src/io/base_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#include "copc-lib/io/base_reader.hpp"
+
+#include <cmath>
```
Comment (leo-stan) [inline cpp/src/io/base_reader.cpp:3]:
```suggestion

```

File: cpp/src/io/base_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#include <iomanip>
+#include <iostream>
+#include <stdexcept>
+
+#include "copc-lib/copc/copc_config.hpp"
+#include "copc-lib/copc/extents.hpp"
+#include "copc-lib/hierarchy/internal/hierarchy.hpp"
```
Comment (leo-stan) [inline cpp/src/io/base_reader.cpp:10]:
```suggestion

```

File: cpp/src/io/base_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#include "copc-lib/las/header.hpp"
+#include "copc-lib/laz/decompressor.hpp"
```
Comment (leo-stan) [inline cpp/src/io/base_reader.cpp:12]:
```suggestion

```

File: cpp/src/io/base_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+uint64_t BaseReader::FetchVlr(const std::map<uint64_t, las::VlrHeader> &vlrs, const std::string &user_id,
+                              uint16_t record_id)
+{
+    for (auto &[offset, vlr_header] : vlrs)
+    {
+        if (vlr_header.user_id == user_id && vlr_header.record_id == record_id)
+        {
+            return offset;
+        }
```
Comment (leo-stan) [inline cpp/src/io/base_reader.cpp:98]:
```suggestion
        if (vlr_header.user_id == user_id && vlr_header.record_id == record_id)
            return offset;
```

File: cpp/include/copc-lib/io/base_reader.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  public:
+    BaseReader(std::istream *in_stream) : in_stream_(in_stream) { InitReader(); }
+
+  protected:
+    BaseReader() = default;
+
+    las::LazConfig las_config_;
+    std::map<uint64_t, las::VlrHeader> vlrs_; // maps from absolute offsets to VLR entries
+
+    std::istream *in_stream_;
```
Comment (leo-stan) [inline cpp/include/copc-lib/io/base_reader.hpp:26]:
```suggestion
    std::istream *in_stream_{};
```

Comment (leo-stan) [reply cpp/include/copc-lib/io/base_reader.hpp:26]: Is it better practice to use a unique_ptr instead?

File: cpp/include/copc-lib/io/laz_reader.hpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
+#ifndef COPCLIB_IO_LAZ_READER_H_
+#define COPCLIB_IO_LAZ_READER_H_
+
+#include "copc-lib/io/base_reader.hpp"
+
+namespace copc::laz
```
Comment (leo-stan) [inline cpp/include/copc-lib/io/laz_reader.hpp:6]: Looks like we use copc::las for the config stuff and copc::laz for the reader/writer haha pretty confusing, then the config is `LazConfig`, but then the header is `LasHeader`, I wonder if we should just put it all as laz/Laz? Less stuff to remember :smile:

File: cpp/include/copc-lib/copc/copc_config.hpp [hunk 7]
```diff
@@ -18,11 +18,16 @@ class CopcConfig : public las::LazConfig
 {
   public:
     CopcConfig() = default;
+
     CopcConfig(const las::LasHeader &header, const CopcInfo &copc_info, const CopcExtents &copc_extents,
                const std::string &wkt, const las::EbVlr &extra_bytes_vlr)
         : LazConfig(header, wkt, extra_bytes_vlr), copc_info_(std::make_shared<copc::CopcInfo>(copc_info)),
           copc_extents_(std::make_shared<copc::CopcExtents>(copc_extents)){};
 
+    CopcConfig(const LazConfig &laz_config, const CopcInfo &copc_info, const CopcExtents &copc_extents)
```
Comment (leo-stan) [inline cpp/include/copc-lib/copc/copc_config.hpp:27]: Should we test this constructor explicitly in `copc_config_test.cpp`?

File: cpp/src/io/copc_reader.cpp [hunk 8]
```diff
@@ -16,67 +16,20 @@
 namespace copc
```
Comment (leo-stan) [inline cpp/src/io/copc_reader.cpp:16]: There are a couple unused includes

**Comment by leo-stan** [comment]
I wonder if we should reach out to the laspy guys now that we have a LAZ reader/writer for cpp, they might reference copclib as cpp alternatives

File: cpp/src/io/copc_reader.cpp (hunk 8 above)
Comment (CCInc) [reply cpp/src/io/copc_reader.cpp:16]: I can't detect them in my IDE unfortunately

**Comment by CCInc** [comment]
Eh, it's not fully featured in any way. Just works well enough for me 😄

File: cpp/include/copc-lib/io/base_reader.hpp (hunk 5 above)
Comment (CCInc) [reply cpp/include/copc-lib/io/base_reader.hpp:26]: I don't think we need a pointer at all but removing that would be pretty substantial

File: cpp/src/io/copc_reader.cpp (hunk 8 above)
Comment (leo-stan) [reply cpp/src/io/copc_reader.cpp:16]: Those two
```
#include "copc-lib/las/header.hpp"
#include <lazperf/readers.hpp>
```

**Review by CCInc — APPROVED** [review]

**Comment by hobu** [comment]
> I wonder if we should reach out to the laspy peeps now that we have a LAZ reader/writer for cpp, they might reference copclib as alternative for cpp

Given that the laspy maintainer is the lazrs developer and uses that for laspy now, I doubt there would be much interest. btw, laspy has local/remote COPC read support now, but the write side is of course more complicated and not available in laspy.

**Pull request merged**